li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java

```diff
@@ -610,38 +623,84 @@ public void uncaughtException(Thread t, Throwable e) {
     // TODO: ensureOpen() would throw an exception if consumers are already closed. In this case, we might want to send
     // an error message to notify mario that the config reload command fails due to consumer/producer already closed and
     // let mario decide what to do next (re-send the configs etc).
     ensureOpen();
 
+    Set<TopicPartition> curPartitions = assignment();
+
     closeNoLock(RELOAD_CONFIG_EXECUTION_TIME_OUT.toMillis(), TimeUnit.MILLISECONDS);
 
+    // save the original configs in case re-create consumers with new configs failed, we need to re-create
+    // with the original configs
+    Map<String, Object> originalConfigs = _commonConsumerConfigs.originals();
+
     // update existing configs with newConfigs
     // originals() would return a copy of the internal consumer configs, put in new configs and update existing configs
     Map<String, Object> configMap = _commonConsumerConfigs.originals();
     configMap.putAll(newConfigs);
     _commonConsumerConfigs = new LiKafkaConsumerConfig(configMap);
 
-    // reset the state of _consumers and _closed so that subsequent calls will try to instantiate the consumers with new configs
-    _consumers.clear();
+    // re-create per-cluster consumers with new set of configs
+    // if any error occurs, recreate all per-cluster consumers with previous last-known-good configs and
+    // TODO : send an error back to Mario
+    // _consumers should be filled when reload config happens
+    List<ClusterConsumerPair<K, V>> newConsumers = new ArrayList<>();
+
+    try {
+      for (ClusterConsumerPair<K, V> entry : _consumers) {
+        LiKafkaConsumer<K, V> curConsumer = createPerClusterConsumer(entry.getCluster());
+        newConsumers.add(new ClusterConsumerPair<K, V>(entry.getCluster(), curConsumer));
+      }
+
+      // replace _consumers with newly created consumers
+      _consumers.clear();
+      _consumers = newConsumers;
+    } catch (Exception e) {
+      // if any exception occurs, re-create per-cluster consumers with last-known-good configs
+      LOG.error("Failed to recreate per-cluster consumers with new config with exception, restore to previous consumers ", e);
+
+      // recreate failed, restore to previous configured consumers
+      newConsumers.clear();
+      _commonConsumerConfigs = new LiKafkaConsumerConfig(originalConfigs);
+
+      for (ClusterConsumerPair<K, V> entry : _consumers) {
+        LiKafkaConsumer<K, V> curConsumer = createPerClusterConsumer(entry.getCluster());
+        newConsumers.add(new ClusterConsumerPair<K, V>(entry.getCluster(), curConsumer));
+      }
+
+      _consumers = newConsumers;
+    }
+
     _closed = false;
 
+    // keep previous partition assignment after config reload
+    // TODO: after subscribe() is supported, we need to also keep the original subscription
+    Map<ClusterDescriptor, Set<TopicPartition>> perClusterTopicPartitionSet = getPerClusterTopicPartitionSet(curPartitions);
```

Comment:
Similar comment as above. Why don't you use a map from cluster descriptor to topic partition set to begin with by calling per-cluster consumer's assignment()?